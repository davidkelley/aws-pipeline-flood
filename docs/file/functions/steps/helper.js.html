<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <base data-ice="baseUrl" href="../../../">
  <title data-ice="title">functions/steps/helper.js | Event Streamer Service</title>
  <link type="text/css" rel="stylesheet" href="css/style.css">
  <link type="text/css" rel="stylesheet" href="css/prettify-tomorrow.css">
  <script src="script/prettify/prettify.js"></script>
  <script src="script/manual.js"></script>
<meta name="description" content="Pushes CodePipeline, CodeBuild, Build Logs and CloudFormation updates to IoT"><meta property="og:type" content="website"><meta property="og:url" content="http://unwrapped.notonthehighstreet.com/aws-pipeline-streamer/"><meta property="og:site_name" content="Event Streamer Service"><meta property="og:title" content="Event Streamer Service"><meta property="og:image" content="https://cdn.notonthehighstreet.com/images/header/logos/tag_logo_uk@2x.png"><meta property="og:description" content="Pushes CodePipeline, CodeBuild, Build Logs and CloudFormation updates to IoT"><meta property="og:author" content="https://twitter.com/notonthehighst"><meta property="twitter:card" content="summary"><meta property="twitter:title" content="Event Streamer Service"><meta property="twitter:description" content="Pushes CodePipeline, CodeBuild, Build Logs and CloudFormation updates to IoT"><meta property="twitter:image" content="https://cdn.notonthehighstreet.com/images/header/logos/tag_logo_uk@2x.png"></head>
<body class="layout-container" data-ice="rootContainer">

<header>
  <a href="./" style="display: flex; align-items: center;"><img src="./image/brand_logo.png" style="width:34px;"></a>
  
  <a href="identifiers.html">Reference</a>
  <a href="source.html">Source</a>
  <a href="test.html" data-ice="testLink">Test</a>
  <div class="search-box">
  <span>
    <img src="./image/search.png">
    <span class="search-input-edge"></span><input class="search-input"><span class="search-input-edge"></span>
  </span>
    <ul class="search-result"></ul>
  </div>
<a style="position:relative; top:3px;" href="https://github.com/notonthehighstreet/aws-pipeline-streamer"><img width="20px" src="./image/github.png"></a></header>

<nav class="navigation" data-ice="nav"><div>
  <ul>
    
  <li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-AWS_REGION">AWS_REGION</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-FLOODS_PATH">FLOODS_PATH</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-FLOOD_API_KEY">FLOOD_API_KEY</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-FLOOD_API_URI">FLOOD_API_URI</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-GRIDS_PATH">GRIDS_PATH</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-LOG_LEVEL">LOG_LEVEL</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-external">E</span><span data-ice="name"><span><a href="https://github.com/trentm/node-bunyan">Logger</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#pipeline">pipeline</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-handler">handler</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#pipeline-pipeline">pipeline/pipeline</a><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/functions/pipeline/pipeline/artifact.js~Artifact.html">Artifact</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/functions/pipeline/pipeline/resume.js~Resume.html">Resume</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/functions/pipeline/pipeline/start.js~Start.html">Start</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-FILE_BUCKET">FILE_BUCKET</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-RUNNING">RUNNING</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-STATE_MACHINE">STATE_MACHINE</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-SUCCEEDED">SUCCEEDED</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#pipeline-pipeline-parameters">pipeline/pipeline/parameters</a><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/functions/pipeline/pipeline/parameters/parameters.js~Parameters.html">Parameters</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#pipeline-pipeline-uploader">pipeline/pipeline/uploader</a><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/functions/pipeline/pipeline/uploader/file.js~File.html">File</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/functions/pipeline/pipeline/uploader/uploader.js~Uploader.html">Uploader</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#pipeline-pipeline-validate">pipeline/pipeline/validate</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-validate">validate</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-Schema">Schema</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#steps">steps</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-parseFloodResponse">parseFloodResponse</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-transformObjectProps">transformObjectProps</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#steps-create">steps/create</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-handler">handler</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#steps-remove">steps/remove</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-handler">handler</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#steps-results">steps/results</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-handler">handler</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#steps-status">steps/status</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-handler">handler</a></span></span></li>
</ul>
</div>
</nav>

<div class="content" data-ice="content"><h1 data-ice="title">functions/steps/helper.js</h1>
<pre class="source-code line-number raw-source-code"><code class="prettyprint linenums" data-ice="content">/**
 * Attempts to parse a flood.io response, extracting the flood uuid, the grid uuid and the status.
 *
 * @param {Object} data - response from flood io
 * @return {Object} - an object with the flood, grid and status properties
 * ```
 * {
 *  flood: {
 *    uuid
 *  },
 *  grid: {
 *    uuid
 *  },
 *  status
 * }
 * ```
 */
export const parseFloodResponse = (data) =&gt; {
  const { uuid, _embedded, status } = data;
  const grid = _embedded.grids.shift();
  return {
    flood: {
      uuid,
    },
    grid: {
      uuid: grid.uuid,
    },
    status,
  };
};


const makeArrayKey = key =&gt; {
  if (key.length &gt; 2 &amp;&amp; key.lastIndexOf(&apos;[]&apos;) === key.length - 2) {
    return key;
  }
  return `${key}[]`;
};

/**
 * Attempts to remap an object structure to a form compatible object.
 * Takes an object like
 * ```
 * {
 *  flood: {
 *    tool: &apos;jmeter&apos;
 *    grids: [
 *      {
 *        region: &apos;eu-west-1&apos;
 *      }
 *    ]
 *  }
 * }
 * ```
 * and transforms it to
 * ```
 * {
 *  &apos;flood[tool]&apos;: &apos;jmeter&apos;,
 *  &apos;flood[grids][][region]&apos;: &apos;eu-west-1&apos;
 * }
 * ```
 *
 * @param {Object} obj - object containing the flood io preferences
 * @param {Object} parent - parent object to insert into (used for recursion)
 * @param {String} pre - prefix, will be generated from keys if not provided
 */
export const transformObjectProps = (obj, parent, pre) =&gt; {
  const mappedObj = parent || {};

  Object.keys(obj).forEach(prop =&gt; {
    const key = pre ? (`${pre}[${prop}]`) : prop;

    if (typeof obj[prop] === &apos;object&apos; &amp;&amp; !Array.isArray(obj[prop])) {
      transformObjectProps(obj[prop], mappedObj, key);
    } else if (Array.isArray(obj[prop])) {
      obj[prop].forEach(value =&gt; {
        const arrayKey = makeArrayKey(key);

        if (typeof value === &apos;object&apos;) {
          transformObjectProps(value, mappedObj, arrayKey);
        } else {
          mappedObj[arrayKey] = value;
        }
      });
    } else {
      mappedObj[key] = obj[prop];
    }
  });

  return mappedObj;
};
</code></pre>

</div>

<footer class="footer">
  Generated by <a href="https://esdoc.org">ESDoc<span data-ice="esdocVersion">(1.0.4)</span><img src="./image/esdoc-logo-mini-black.png"></a>
</footer>

<script src="script/search_index.js"></script>
<script src="script/search.js"></script>
<script src="script/pretty-print.js"></script>
<script src="script/inherited-summary.js"></script>
<script src="script/test-summary.js"></script>
<script src="script/inner-link.js"></script>
<script src="script/patch-for-local.js"></script>
</body>
</html>
