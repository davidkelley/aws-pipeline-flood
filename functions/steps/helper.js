/**
 * Attempts to parse a flood.io response, extracting the flood uuid, the grid uuid and the status.
 *
 * @param {Object} data - response from flood io
 * @return {Object} - an object with the flood, grid and status properties
 * ```
 * {
 *  flood: {
 *    uuid
 *  },
 *  grid: {
 *    uuid
 *  },
 *  status
 * }
 * ```
 */
export const parseFloodResponse = (data) => {
  const { uuid, _embedded, status } = data;
  const grid = _embedded.grids.shift();
  return {
    flood: {
      uuid,
    },
    grid: {
      uuid: grid.uuid,
    },
    status,
  };
};


const makeArrayKey = key => {
  if (key.length > 2 && key.lastIndexOf('[]') === key.length - 2) {
    return key;
  }
  return `${key}[]`;
};

/**
 * Attempts to remap an object structure to a form compatible object.
 * Takes an object like
 * ```
 * {
 *  flood: {
 *    tool: 'jmeter'
 *    grids: [
 *      {
 *        region: 'eu-west-1'
 *      }
 *    ]
 *  }
 * }
 * ```
 * and transforms it to
 * ```
 * {
 *  'flood[tool]': 'jmeter',
 *  'flood[grids][][region]': 'eu-west-1'
 * }
 * ```
 *
 * @param {Object} obj - object containing the flood io preferences
 * @param {Object} parent - parent object to insert into (used for recursion)
 * @param {String} pre - prefix, will be generated from keys if not provided
 */
export const transformObjectProps = (obj, parent, pre) => {
  const mappedObj = parent || {};

  Object.keys(obj).forEach(prop => {
    const key = pre ? (`${pre}[${prop}]`) : prop;

    if (typeof obj[prop] === 'object' && !Array.isArray(obj[prop])) {
      transformObjectProps(obj[prop], mappedObj, key);
    } else if (Array.isArray(obj[prop])) {
      obj[prop].forEach(value => {
        const arrayKey = makeArrayKey(key);

        if (typeof value === 'object') {
          transformObjectProps(value, mappedObj, arrayKey);
        } else {
          mappedObj[arrayKey] = value;
        }
      });
    } else {
      mappedObj[key] = obj[prop];
    }
  });

  return mappedObj;
};
