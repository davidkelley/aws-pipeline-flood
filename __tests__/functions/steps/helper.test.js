/* eslint-disable import/no-unresolved */

import { parseFloodResponse, transformObjectProps } from '@functions/steps/helper';
import mockResponse from './mock/mockresponse.json';

const mockFloodProps = {
  flood: {
    tool: 'jmeter',
    grids: [
      {
        region: 'eu-west-1',
        instance_type: 'm4.xlarge',
        stop_after: '60',
        infrastructure: 'demand',
      },
    ],
  },
};

describe('Helper suite', () => {
  describe('parseFloodResponse', () => {
    it('Extracts the correct props and returns a new object', () => {
      expect(parseFloodResponse(mockResponse)).toMatchSnapshot();
    });
  });

  describe('transformObjectProps', () => {
    it('Remaps the object properties to match a form', () => {
      expect(transformObjectProps(mockFloodProps)).toMatchSnapshot();
    });
  });
});
