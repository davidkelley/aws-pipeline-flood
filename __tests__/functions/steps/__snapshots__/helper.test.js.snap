// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper suite parseFloodResponse Extracts the correct props and returns a new object 1`] = `
Object {
  "flood": Object {
    "uuid": "VCFstUfl0AMdtaKYBo5d7A",
  },
  "grid": Object {
    "uuid": "hg8ZjZcifgsuQvDfcnBONw",
  },
  "status": "queued",
}
`;

exports[`Helper suite transformObjectProps Remaps the object properties to match a form 1`] = `
Object {
  "flood[grids][][infrastructure]": "demand",
  "flood[grids][][instance_type]": "m4.xlarge",
  "flood[grids][][region]": "eu-west-1",
  "flood[grids][][stop_after]": "60",
  "flood[tool]": "jmeter",
}
`;
